File: apps/product_management/views/product_area_admin.py
from django.http import JsonResponse
from django.views.decorators.http import require_POST
from django.contrib.admin.views.decorators import staff_member_required
from ..models import ProductArea

@staff_member_required
def get_nodes(request):
    exclude_id = request.GET.get('exclude')
    nodes = ProductArea.objects.exclude(id=exclude_id).select_related('parent')
    return JsonResponse([{'id': node.id, 'name': node.name, 'depth': node.get_depth()} for node in nodes], safe=False)

@require_POST
@staff_member_required
def move_node(request, node_id):
    node = ProductArea.objects.get(id=node_id)
    parent_id = request.POST.get('parent')
    if parent_id:
        parent = ProductArea.objects.get(id=parent_id)
        node.move(parent)
    else:
        node.move(None)  # Move to root
    return JsonResponse({'status': 'success'})

File: apps/static/admin/js/product_area_admin.js
(function($) {
    $(document).ready(function() {
        var $moveForm = $('<form id="move-node-form">' +
            '<select name="parent" id="parent-select">' +
            '<option value="">Move to root</option>' +
            '</select>' +
            '<input type="submit" value="Move">' +
            '</form>');

        $('.move_node').click(function(e) {
            e.preventDefault();
            var $row = $(this).closest('tr');
            var nodeId = $row.find('input[name="_selected_action"]').val();

            // Populate select with all nodes except the current one and its descendants
            $.get('/admin/product_management/productarea/get_nodes/?exclude=' + nodeId, function(data) {
                $('#parent-select').html('<option value="">Move to root</option>');
                data.forEach(function(node) {
                    $('#parent-select').append($('<option>', {
                        value: node.id,
                        text: '—'.repeat(node.depth) + ' ' + node.name
                    }));
                });
            });

            $moveForm.dialog({
                title: 'Move Node',
                modal: true,
                buttons: {
                    "Move": function() {
                        var parentId = $('#parent-select').val();
                        $.post('/admin/product_management/productarea/' + nodeId + '/move/', 
                               {parent: parentId, csrfmiddlewaretoken: $('input[name="csrfmiddlewaretoken"]').val()},
                               function(data) {
                                   location.reload();
                               });
                        $(this).dialog("close");
                    },
                    "Cancel": function() {
                        $(this).dialog("close");
                    }
                }
            });
        });

        $('body').append($moveForm);
    });
})(django.jQuery);

